File: scope-css.js
const fs = require('fs');
const path = require('path');

// Read the CSS file
const cssPath = path.join(__dirname, 'src', 'modal', 'modal.css');
const css = fs.readFileSync(cssPath, 'utf8');

// Add CSS containment at the beginning
const containmentCSS = `/* Rubric Grader Modal Styles */

/* CSS Containment and Isolation */
#rubric-grader-modal {
  contain: layout style paint;
  isolation: isolate;
}

#rubric-grader-modal *,
#rubric-grader-modal *::before,
#rubric-grader-modal *::after {
  box-sizing: border-box !important;
}

`;

// Function to scope a selector
function scopeSelector(selector) {
  // Skip if already scoped
  if (selector.trim().startsWith('#rubric-grader-modal')) {
    return selector;
  }

  // Skip @media, @keyframes, etc.
  if (selector.trim().startsWith('@')) {
    return selector;
  }

  // Handle multiple selectors separated by commas
  const selectors = selector.split(',').map(s => s.trim());
  const scopedSelectors = selectors.map(s => {
    // Skip if it's a pseudo-element or pseudo-class only
    if (s.startsWith(':')) {
      return s;
    }

    // Add scope prefix
    return `#rubric-grader-modal ${s}`;
  });

  return scopedSelectors.join(',\n');
}

// Process CSS
const lines = css.split('\n');
let result = containmentCSS;
let inMediaQuery = false;
let mediaQueryDepth = 0;

for (let i = 0; i < lines.length; i++) {
  const line = lines[i];
  const trimmed = line.trim();

  // Skip the first comment line (already added)
  if (i === 0 && trimmed.startsWith('/*')) {
    continue;
  }

  // Track media queries
  if (trimmed.startsWith('@media')) {
    inMediaQuery = true;
    mediaQueryDepth = 0;
    result += line + '\n';
    continue;
  }

  if (inMediaQuery) {
    result += line + '\n';
    if (trimmed.includes('{')) mediaQueryDepth++;
    if (trimmed.includes('}')) {
      mediaQueryDepth--;
      if (mediaQueryDepth === 0) inMediaQuery = false;
    }
    continue;
  }

  // Check if this line contains a selector (ends with {)
  if (trimmed.endsWith('{') && !trimmed.startsWith('/*') && !trimmed.startsWith('*')) {
    const selector = trimmed.slice(0, -1).trim();
    const scopedSelector = scopeSelector(selector);
    const indent = line.match(/^\s*/)[0];
    result += `${indent}${scopedSelector} {\n`;
  } else {
    result += line + '\n';
  }
}

// Write the scoped CSS
fs.writeFileSync(cssPath, result, 'utf8');
console.log('CSS scoping completed!');
console.log(`Processed ${lines.length} lines`);

File: scope-css.py
import re

# Read the CSS file
with open('src/modal/modal.css', 'r', encoding='utf-8') as f:
    css = f.read()

# Split into lines
lines = css.split('\n')

# Process each line
result_lines = []
in_media_query = False
media_depth = 0
skip_next_scope = False

for i, line in enumerate(lines):
    stripped = line.strip()

    # Skip lines that are already scoped or are special cases
    if i < 25:  # Skip the header we just added
        result_lines.append(line)
        continue

    # Track media queries
    if stripped.startswith('@media'):
        in_media_query = True
        media_depth = 0
        result_lines.append(line)
        continue

    if in_media_query:
        result_lines.append(line)
        if '{' in stripped:
            media_depth += stripped.count('{')
        if '}' in stripped:
            media_depth -= stripped.count('}')
            if media_depth <= 0:
                in_media_query = False
        continue

    # Check if line contains a selector (ends with {)
    if stripped.endswith('{') and not stripped.startswith('/*') and not stripped.startswith('*') and not stripped.startswith('@'):
        # Extract selector
        selector = stripped[:-1].strip()

        # Skip if already scoped
        if selector.startswith('#rubric-grader-modal'):
            result_lines.append(line)
            continue

        # Get indentation
        indent = line[:len(line) - len(line.lstrip())]

        # Scope the selector
        if ',' in selector:
            # Handle multiple selectors
            selectors = [s.strip() for s in selector.split(',')]
            scoped = [f'#rubric-grader-modal {s}' for s in selectors]
            result_lines.append(f"{indent}{',\\n{indent}'.join(scoped)} {{")
        else:
            result_lines.append(f"{indent}#rubric-grader-modal {selector} {{")
    else:
        result_lines.append(line)

# Write back
with open('src/modal/modal.css', 'w', encoding='utf-8') as f:
    f.write('\n'.join(result_lines))

print(f"Processed {len(lines)} lines")
print("CSS scoping completed!")
